Classes/Plugin/MoreLikeThis.php

```diff
@@ -44,96 +44,56 @@ class MoreLikeThis extends PluginBase
+                '\ApacheSolrForTypo3\Solrmlt\Configuration\PluginConfiguration',
+                $this);
+        }
 
-        return $actionResult;
+        return $this->pluginConfiguration;
     }
 
     /**
-     * This method is used to create a moreLikeThis query and
-     * initializes it with the needed values.
-     *
-     * @return MoreLikeThisQuery
+     * @return \ApacheSolrForTypo3\Solrmlt\Query\Builder
      */
-    protected function getMoreLikeThisQuery()
+    protected function getQueryBuilder()
     {
-        $query = null;
-
-        if ($this->solrAvailable) {
-            $query = GeneralUtility::makeInstance('ApacheSolrForTypo3\\Solrmlt\\Query\\MoreLikeThisQuery');
-
-            $query->setUserAccessGroups(explode(',', $GLOBALS['TSFE']->gr_list));
-            $query->setSiteHashFilter(Site::getSiteByPageId($GLOBALS['TSFE']->id)->getDomain());
-
-            $query->setQueryFields(array('id,title', 'score'));
-
-            $query->setSimilarityFields(GeneralUtility::trimExplode(
-                ',',
-                $this->pi_getFFvalue(
-                    $this->cObj->data['pi_flexform'], 'similarityFields'
-                ),
-                true
-            ));
-            $query->setMinimumTermFrequency($this->pi_getFFvalue(
-                $this->cObj->data['pi_flexform'], 'minTermFrequency', 'sAdvanced'
-            ));
-            $query->setMinimumDocumentFrequency($this->pi_getFFvalue(
-                $this->cObj->data['pi_flexform'], 'minDocumentFrequency', 'sAdvanced'
-            ));
-            $query->setMinimumWordLength($this->pi_getFFvalue(
-                $this->cObj->data['pi_flexform'], 'minWordLength', 'sAdvanced'
-            ));
-            $query->setMaximumWordLength($this->pi_getFFvalue(
-                $this->cObj->data['pi_flexform'], 'maxWordLength', 'sAdvanced'
-            ));
-            $query->setMaximumQueryTerms($this->pi_getFFvalue(
-                $this->cObj->data['pi_flexform'], 'maxQueryTerms', 'sAdvanced'
-            ));
+        if ($this->queryBuilder == null) {
+            $this->queryBuilder = GeneralUtility::makeInstance('ApacheSolrForTypo3\Solrmlt\Query\Builder');
         }
 
-        return $query;
+        return $this->queryBuilder;
     }
 
     /**
-     * Determines the moreLikeThis Id string to be used for the MLT query.
+     * Creates a moreLikeThis query and returns the Apache_Solr_Response for the
+     * query. The response is processed in the render method.
      *
-     * @return string The string to be used to find similar documents
+     * @see classes/pluginbase/tx_solr_pluginbase_PluginBase#performAction()
+     * @return \Apache_Solr_Response The Solr server's response
      */
-    protected function getMoreLikeThisIdString()
+    protected function performAction()
     {
+        $query = $this->getQueryBuilder()->build($this->getPluginConfiguration(), $GLOBALS['TSFE']);
```

Comment:
probably should also add a wrapper getTSFE() to easily provide a faked TSFE when testing